formats/utils/color-factory.js

```diff
@@ -0,0 +1,47 @@
+/* eslint-disable id-length */
+const tinycolor = require('tinycolor2');
+const {hexToHsluv, hsluvToRgb} = require('hsluv');
+
+const colorFactory = (colors, roleVariants, colorScheme) => {
+  return Object.assign(
+    {},
+    ...Object.entries(colors).map(([role, hex]) => {
+      const base = hexToHsluv(hex);
+      const variants = roleVariants[role] || [];
+      return {
+        ...variants.reduce((accumulator, {name, ...settings}) => {
+          const {
+            hue = base[0],
+            saturation = base[1],
+            lightness = base[2],
+            alpha = 1,
+          } = settings[colorScheme];
+
+          const resolve = (value, baseToResolve) => {
+            return typeof value === 'number' ? value : value(baseToResolve);
+          };
+
+          const [red, green, blue] = hsluvToRgb([
+            resolve(hue, base[0]),
+            resolve(saturation, base[1]),
+            resolve(lightness, base[2]),
+          ]);
+
+          return {
+            ...accumulator,
+            [name]: tinycolor({
```

Comment:
I don't know if we should use `tinycolor` here because it will become a dependency (instead of dev dependency) and then become part of the Polaris React bundle. Maybe it is worth copying over the few color functions we need or maybe tinycolor is small enough?

Reply:
That's a good point, but we get so much usefulness out of including it, and it really isn't that big (https://bundlephobia.com/result?p=tinycolor2@1.4.1). The built in functionality for accessibility checking could be tremendously powerful. it would also allow us to eventually remove all the color-related code from Polaris React, which would offset some of the weight gain.

I'm on the fence with this decision.

Reply:
Another option still would be to use [`colors.js`](https://bundlephobia.com/result?p=colors.js@1.2.4) for this simple color conversion, which would also do away with the color code we're currently maintaining in Polaris React. Then we could use tinycolor in the test environment for contrast checking here in tokens.

Reply:
Or remove everything and just use the `hsluvToRgb` function 😄